### DISCUSSION TITLE
job that has been killed for exceeding "shutdown-timeout" has inconsistent display

### DISCUSSION THREAD

**Discussion by jrochkind** [bensheldon/good_job#650] [Q&A]
I set up a local demonstration of the `shutdown-timeout` feature, to make sure I understood what happened to a job here. Mostly it looks good! But there is one weird part. 

I had a long-running job, and I sent a SIGTERM to `good_job start` process.  After `shutdown-timeout` was exceeded, the good_job process quits great. 

Looking in the good job dashboard, what happens to that job that got interupted? It shows up in the _tab_ "Quened", but with the "state" "Running". 

![Screen Shot 2022-06-30 at 2 49 32 PM](https://user-images.githubusercontent.com/149304/176755193-cc45e67d-70c1-44bc-97ba-4e335197eff3.png)

It's definitely not truly running, there isn't even a `good_job` worker process running. And it is in tab "queued", I'd consider it queued. But it has state "running"? That seems to conflict with it being in tab "queued" though?  And if you look at tab "All", it just shows up with state "running", you wouldn't know it's really actually in a sort of queued state. 

(If I start up `good_job` again, a worker *does* dequeue that job to work on, that works properly!  Although no matter how many times happens, it's "executions" count stays at `1`). 

Is all this expected, or behaving in an unexpected way?  Should it be possible for a job to have state "running" but be in tab "queued"?

### Also, there's no way to cancel it

This interupted job, all the action buttons in dashboard are greyed out. What if I actually do want to discard it now, when it's in this holding state? I can't. Expected?

### Also, logging it?

If I look at stdout in my console, the fact that it shutdown is logged (kind of... twice?)

```
[GoodJob] [60566] GoodJob shutting down scheduler...
[GoodJob] [60566] GoodJob scheduler is shut down.
[GoodJob] Notifier unsubscribed with UNLISTEN
[GoodJob] [60566] GoodJob shutting down scheduler...
[GoodJob] [60566] GoodJob scheduler is shut down.
[GoodJob] [60566] [GoodJob::Scheduler(queues=* max_threads=5)-thread-3] Executed GoodJob f02e8124-088d-418c-925a-01c4a20478d6
```

But it doesn't log that it's waiting up to `shutdown-timeout` for jobs to complete. Should it?  

It also doesn't log that jobs were interupted; would it be feasible for it to log the job details of interrupted jobs?  This is an unusual occurence it might be nice to have in the logs to aid with debugging unusual occurences.

**Comment by bensheldon** [comment]
Thanks for flagging the inconsistency! I think I know what's going on, but haven't dug myself into the code.

The most-valid definition of "running" is "a connection holds an advisory lock on the job". Querying advisory locks has a cost, and also some complexity to avoid N+1 queries checking for locks in the dashboard.

There is secondarily a column for `good_jobs.performed_at` that is touched when a job is picked up for execution. This is only used for display/metrics; the job executor solely uses advisory locks. 

When you forced the shutdown, the Advisory Lock was released, but the `good_jobs.perform_at` value is still set. The Dashboard _tries_ to be smart about querying for Advisory Locks rather than the `performed_at` value, but something must be off about it. That's a dashboard bug 🐞 Nice find! 

### Logging

Those double logs of Scheduler shutdown is odd... I think. If you are using multiple thread-pools (e.g. `GOOD_JOB_QUEUES=queue1:3;queue2:3") or (rare) you have assigned async adapters to multiple Job classes directly, you can end up with multiple schedulers in the same process. 

I do think we could do better logging. Something like:

```
GoodJob scheduler waiting 30 seconds for 3 jobs to finish before shutting down
GoodJob scheduler forcibly interrupting 1 unfinished job to shutdown
GoodJob scheduler is shut down
```

Here's where those notification hooks live:

https://github.com/bensheldon/good_job/blob/b86000930f1ddb6d34cb3ad4c65214dea00e6a30/lib/good_job/scheduler.rb#L112-L127

**Reply by jrochkind** [reply]
Thanks!

i'm not doing anything weird with my config as far as I know. 

In logging, I'd love to see the ActiveJob ID or other identifiers of any forcibly interupted jobs... if that information is accessible at the point of logging and that's possible.  Then if I'm trying ot figure out "wait, what weird things are going on with THIS job", that event is captured in logs and findable by looking up job-id(s). 

I'm not sure if any "cleanup" is done for interupted jobs, or if they are just... killed, the end. If there's some cleanup, that'd be a good place to capture their IDs for logging, and also a good place to clean up any inconsistent database values that may be leading to the inconsistent dashboard reporting. 

Sidekiq an Resque both do run some cleanup (including persisting things) on interupting a task -- that does mean it may take a _bit_ longer than your shutdown-timeout for the thing to really exit, as it performs cleanup. That is fine, it just means that for instance on heroku you set the sidekiq/resque equivalent of shutdown-timeout to be well under heroku's real timeout, to leave some time for the cleanup.

**Reply by jrochkind** [reply]
The double "shutting down schedule... scheduler is shut down" messages happen only when there is a "running" job in progress when SIGTERM is received.  I wonder if someone else can reproduce it just by shutting down with an in-progress running job that has to be waited for. 

Perhaps I'll have time/interest to debug it at some point, not sure.

**Reply by bensheldon** [reply]
Oh! That extra details makes me think I know what's causing it. When there is a shutdown with timeout, GoodJob has a helper method that calls shutdown across all shutdown-able objects twice:

https://github.com/bensheldon/good_job/blob/275c372604973ff767ada7ebc409b4934a472526/lib/good_job.rb#L117-L126

First it sends a shutdown without a timeout to everything, and then it comes back and sends another shutdown to each one this time with a timeout. 

I think this could be addressed by being clearer in the `shutdown` instrumentation what was asked of for e.g. "started the shutdown" vs "shutting down and waiting for it to shutdown" and then the end state of "ok, it's actually been shut down"

**Reply by jrochkind** [reply]
I went to look at improving the shutdown logging...

What I _really_ want is to log the details (such as JobID) of any jobs that are forcefully quit in progress, after a timeout.  Because this is a pretty "traumatic" thing for a job that might be useful in debugging "Wait, what happened to these jobs?"  But there's actually very little observability into it -- I don't think you can tell from the dashboard that this has happened to a job. I am not sure if you can tell by looking at anything in the job database row. 

However, it doesn't look feasible to do this logging without changes to GoodJob. The "hard" shutdown of in-progress jobs is done by [sending #kill to a ThreadPoolExeuctor](https://github.com/bensheldon/good_job/blob/a8214cee1c3554383b59e83527f1251547de73b8/lib/good_job/scheduler.rb#L124), and letting it handle it. There's really no way to look "inside" the ThreadPoolExeuctor to find out anything about the threads it's currently excecuting, the ruby-concurrent API doesn't work that way. So there doesn't seem to be a way to log details on the jobs that are interupted this way. 

Next I tried to look at just improving logging of the shutdown so the two shutdowns at least aren't logged identically. Although I'm not stotally sure what you'd call them. The first one is like "shutdown attempt, with wait", and the second is like "shut down right now interupting processes". But I didn't understand the way the instrumentation/logging system is working, and other context, well enough to make that change.  And while it does seem cleaner to have these separated in log messages, I'm not totally sure how useful it actually is in practice (not having the practice yet), so chose not to spend more time on it for now, sorry.

**Reply by jrochkind** [reply]
on the logging of interupted jobs... wait _something_ changes the job status such that GoodJob knows to re-run the job when it gets interupted. So there must be _some_ place that could catch it getting interrupted and log it... I am not sure where it is though. 

I would love to see that interuption logged; it's actually not crucial if it knows _why_ it was interupted (an intentional shutdown after timeout, or something more catastrophic), but it would be great if it were logged with JobID details that it got interupted.

**Reply by bensheldon** [reply]
Agreed, it's tough to know which jobs in which threads are getting killed for the reasons you cite. 

> wait something changes the job status such that GoodJob knows to re-run the job

The presence/absence of an Advisory Lock is what locks the job, and the (generally a strength) about them is they are _implicitly_ released when the database connection disconnects. There's not an "event" to hook into in Ruby.

I had an idea to add a `process_id` column to the `good_jobs` table and write the `GoodJob::Process.current_id` into it. That would make it queryable, but also emitting database queryies when we're trying to shutdown the process is maybe not great.

You can query for jobs that have been killed (assuming they haven't been re-executed): `GoodJob::Execution.where(finished_at: nil).where.not(performed_at: nil).advisory_unlocked` e.g. `performed_at` is present, `finished_at` is absent, but the record doesn't have an active Advisory Lock

What about this?

When the job is executed, it updates the Thread name (something like [this](https://github.com/bensheldon/good_job/blob/26d16402fd2f3a81de7851dc07287de87e9180ac/lib/good_job/scheduler.rb#L268-L268). And then log the thread names that are active before telling the Scheduler to hard terminate them. I _think_ we can get the Thread names out of the Scheduler, but that might require some digging in Concurrent::Ruby.

**Reply by bensheldon** [reply]
@jrochkind I had a burst of insight and come up with an approach to capture the jobs that are killed and log them.

Could you give me feedback on this log message? https://github.com/bensheldon/good_job/pull/794/files#diff-3ba7e5bf9d360520b1c356c89199f8cd105b396c0a94fb3fe56a34abe12ddb2bR104

**Reply by jrochkind** [reply]
Ooh, very nice!  That log message seems reasonable to me.  The important thing is that it includes the ActiveJob IDs!

**Comment by Am0rphous** [comment]
Virustotal flags the links in the comment from KimberlyCollins5047 as phishing and suspicious.

The user have been reported.

Stay safe fellas.
